File: cfg_maker/cs_test/gcf/CrossCall/Binary.cs
/*
 * 二进制序列化
 */

using System;

namespace CrossCall
{
    /*
     * 二进制数据Reader 
     */
    public class BinaryReader : Serialize.IReader
    {
        protected IBinaryStream _stream = null;

        public BinaryReader(IBinaryStream stream)
        {
            _stream = stream;
        }

        public bool StructBegin(uint code, string name = null)
        {
            uint check = 0;
            if (!Read(ref check, null))
                return false;

            return check == code;
        }

        public bool StructEnd()
        {
            return true;
        }

        public bool ArrayBegin(ref int length, string name = null)
        {
            byte[] data = { 0, 0, 0, 0 };
            if (!_stream.Read(data)) return false;

            length = BitConverter.ToInt32(data, 0);
            return true;
        }

        public bool ArrayEnd()
        {
            return true;
        }

        public bool Read(ref bool val, string name = null)
        {
            byte[] data = { 0 };
            if (!_stream.Read(data)) return false;

            val = BitConverter.ToBoolean(data, 0);
            return true;
        }

        public bool Read(ref int val, string name = null)
        {
            byte[] data = { 0, 0, 0, 0 };
            if (!_stream.Read(data)) return false;

            val = BitConverter.ToInt32(data, 0);
            return true;
        }

        public bool Read(ref uint val, string name = null)
        {
            byte[] data = { 0, 0, 0, 0 };
            if (!_stream.Read(data)) return false;

            val = BitConverter.ToUInt32(data, 0);
            return true;
        }

        public bool Read(ref float val, string name = null)
        {
            byte[] data = { 0, 0, 0, 0 };
            if (!_stream.Read(data)) return false;

            val = BitConverter.ToSingle(data, 0);
            return true;
        }

        public bool Read(ref string val, string name = null)
        {
            int len = 0;
            if (!Read(ref len, null))
                return false;

            byte[] data = new byte[len];
            if (!_stream.Read(data))
                return false;

            val = System.Text.Encoding.UTF8.GetString(data);
            return true;
        }
    }

    /*
     * 二进制数据Writer
     */
    public class BinaryWriter : Serialize.IWriter
    {
        protected IBinaryStream _stream = null;

        public BinaryWriter(IBinaryStream stream)
        {
            _stream = stream;
        }

        public bool StructBegin(uint code, string name = null)
        {
            return _stream.Write(BitConverter.GetBytes(code));
        }

        public bool StructEnd()
        {
            return true;
        }

        public bool ArrayBegin(int length, string name = null)
        {
            return _stream.Write(BitConverter.GetBytes(length));
        }

        public bool ArrayEnd()
        {
            return true;
        }

        public bool Write(bool val, string name = null)
        {
            return _stream.Write(BitConverter.GetBytes(val));
        }

        public bool Write(int val, string name = null)
        {
            return _stream.Write(BitConverter.GetBytes(val));
        }

        public bool Write(uint val, string name = null)
        {
            return _stream.Write(BitConverter.GetBytes(val));
        }

        public bool Write(float val, string name = null)
        {
            return _stream.Write(BitConverter.GetBytes(val));
        }

        public bool Write(string val, string name = null)
        {
            return _stream.Write(BitConverter.GetBytes(val.Length)) &&
                _stream.Write(System.Text.Encoding.UTF8.GetBytes(val));
        }
    }
}

File: gcf/test/test_cpp/gcf/cross_call/Binary.cpp
#include "Binary.h"
#include <cassert>

namespace cross_call
{
    //////////////////////////////////////////////////////////////////////////
    // NormalReader
    BinaryReader::BinaryReader(IBinaryStream* stream)
        : _stream(stream)
    {
    }

    bool BinaryReader::StructBegin(uint32_t code, const char* name/* = nullptr*/)
    {
        uint32_t check = 0;
        return _stream->Read(&check, 4) && check == code;
    }

    bool BinaryReader::StructEnd()
    {
        return true;
    }

    bool BinaryReader::ArrayBegin(int& length, const char* name /* = nullptr */)
    {
        return _stream->Read(&length, 4);
    }

    bool BinaryReader::ArrayEnd()
    {
        return true;
    }

    bool BinaryReader::Read(bool& val, const char* name/* = nullptr*/)
    {
        return _stream->Read(&val, 1);
    }

    bool BinaryReader::Read(int32_t& val, const char* name/* = nullptr*/)
    {
        return _stream->Read(&val, 4);
    }

    bool BinaryReader::Read(uint32_t& val, const char* name/* = nullptr*/)
    {
        return _stream->Read(&val, 4);
    }

    bool BinaryReader::Read(float& val, const char* name/* = nullptr*/)
    {
        return _stream->Read(&val, 4);
    }

    bool BinaryReader::Read(std::string& val, const char* name/* = nullptr*/)
    {
        uint32_t size = 0;
        char buf[1024] = {0};

        if (!Read(size)) return false;

        assert(size < 1024);
        if (!_stream->Read(buf, size)) return false;

        val = buf;
        return true;
    }

    //////////////////////////////////////////////////////////////////////////
    // NormalWriter
    BinaryWriter::BinaryWriter(IBinaryStream* stream)
        : _stream(stream)
    {
    }

    bool BinaryWriter::StructBegin(uint32_t code, const char* name/* = nullptr*/)
    {
        return _stream->Write(&code, 4);
    }

    bool BinaryWriter::StructEnd()
    {
        return true;
    }

    bool BinaryWriter::ArrayBegin(int length, const char* name /* = nullptr */)
    {
        return _stream->Write(&length, 4);
    }

    bool BinaryWriter::ArrayEnd()
    {
        return true;
    }

    bool BinaryWriter::Write(bool val, const char* name/* = nullptr*/)
    {
        return _stream->Write(&val, 1);
    }

    bool BinaryWriter::Write(int32_t val, const char* name/* = nullptr*/)
    {
        return _stream->Write(&val, 4);
    }

    bool BinaryWriter::Write(uint32_t val, const char* name/* = nullptr*/)
    {
        return _stream->Write(&val, 4);
    }

    bool BinaryWriter::Write(float val, const char* name/* = nullptr*/)
    {
        return _stream->Write(&val, 4);
    }

    bool BinaryWriter::Write(const std::string& val, const char* name/* = nullptr*/)
    {
        uint32_t size = (uint32_t)val.size();
        assert(size < 1024);

        if (!_stream->Write(&size, 4)) return false;

        return _stream->Write(val.c_str(), size);
    }
}
